src/sync.js

```diff
@@ -1294,41 +1294,35 @@ SyncApi.prototype._resolveInvites = function(room) {
  */
 SyncApi.prototype._processRoomEvents = function(room, stateEventList,
                                                 timelineEventList) {
-    timelineEventList = timelineEventList || [];
-    const client = this.client;
-    // "old" and "current" state are the same initially; they
-    // start diverging if the user paginates.
-    // We must deep copy otherwise membership changes in old state
-    // will leak through to current state!
-    const oldStateEvents = utils.map(
-        utils.deepCopy(
-            stateEventList.map(function(mxEvent) {
-                return mxEvent.event;
-            }),
-        ), client.getEventMapper(),
-    );
-    const stateEvents = stateEventList;
+    // We'll only get state events in stateEventList if this is a limited
+    // sync, in which case this should be a fresh timeline, and this is us
+    // initialising it.
 
-    // set the state of the room to as it was before the timeline executes
-    //
-    // XXX: what if we've already seen (some of) the events in the timeline,
-    // and they modify some of the state set in stateEvents? In that case we'll
-    // end up with the state from stateEvents, instead of the more recent state
-    // from the timeline.
-    room.oldState.setStateEvents(oldStateEvents);
-    room.currentState.setStateEvents(stateEvents);
+
+    // If there are no events in the timeline yet, initialise it with
+    // the given state events
+    const timelinewasEmpty = room.getLiveTimeline().getEvents().length == 0;
+    if (timelinewasEmpty) {
+        room.getLiveTimeline().initialiseState(stateEventList);
+    }
 
     this._resolveInvites(room);
```

Comment:
why does this (and recalculate) happen after adding the state events for an empty timeline, but before it for an existing timeline?

Reply:
My reasoning for this was the the state events ought to be treated the same as if they'd arrived in the timeline section. That said, I opted not to chase that particular rabbit down the hole to figure out what actually happens on a room name change, etc.

Reply:
well, if that's true, why isn't it true for when the timeline is empty?

"I didn't go down the rabbithole" is valid but an `XXX` comment saying so would be good for future reference

Reply:
good point, and now looking at it, it looks a bit broken. Have added an XXX to explain how I think it might be broken.